File: 00006.c
// https://leetcode.com/problems/zigzag-conversion/

// minha abordagem usando listas
// feito inicialmente em python, fiz em C para comparar runtime

typedef char KEYTYPE;

typedef struct _ListNode{
	KEYTYPE key;
	struct _ListNode *next;
} ListNode;

typedef struct{
	ListNode *head;
	ListNode *end;
} List;

void init(List *list){
	list->head = NULL;
	list->end = NULL;
}

void insertAtLast(List *list, KEYTYPE value){
	ListNode *head = list->head;
	
	ListNode *newNode = (ListNode *) malloc(sizeof(ListNode));
	newNode->key = value;
	newNode->next = NULL;
	
	if(head == NULL)
		list->head = list->end = newNode;
	else{
		list->end->next = newNode;
		list->end = newNode;
	}
}

void freeList(List *list){
	ListNode *current = list->head;
	while(current != NULL){
		ListNode *remove = current;
		current = current->next;
		free(remove);
	}
	list->head = NULL;
	list->end = NULL;
}

char* convert(char* s, int numRows) {
    int lenght = strlen(s);
    if(lenght <= numRows || numRows == 1)
    	return strdup(s);
    	
    List *rowList = (List *) malloc(numRows * sizeof(List));
    for(int i = 0; i < numRows; ++i)
    	init(rowList+i);
    insertAtLast(rowList, s[0]);
    
	int i = 1; 
    
    while(true){
		int j;
		for(j = 1; j < numRows; ++j){
			insertAtLast(rowList+j, s[i++]);
			if(i == lenght)
                goto ENDLOOP;
		}
		for(j-=2; j >= 0; --j){
			insertAtLast(rowList+j, s[i++]);
			if(i == lenght)
                goto ENDLOOP;
		}
	}
	ENDLOOP: 
	
    char *converted_str = (char *) malloc(lenght+1);
	converted_str[lenght] = '\0';
    
    for(int j = 0, i = 0; j < numRows; ++j){
		ListNode *current = rowList[j].head;
		while(current != NULL){
			converted_str[i++] = current->key;
			current = current->next;
		}
		freeList(rowList+j); // (*)
	}
	
    free(rowList); 
    
    return converted_str;
}

File: 00006.py
# https://leetcode.com/problems/zigzag-conversion

class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1 or len(s) <= numRows:
            return s

        # listas das linhas (a 1a ja começa com s[0]
        rowcharacters = [[s[0]]]
        # inicializa listas vazias
        for i in range(1, numRows):
            rowcharacters.append([])

        i = 1 # caracter atual da string
        finished = False
        
        while True:
            for j in range(1, numRows):
                rowcharacters[j].append(s[i])
                i += 1
                if i == len(s):
                    finished = True
                    break

            if finished: break
            
            j = numRows-2
            while j >= 0:
                rowcharacters[j].append(s[i])
                i += 1
                if i == len(s):
                    finished = True
                    break
                j -= 1

            if finished: break

        for i in range(1, numRows):
            rowcharacters[0].extend(rowcharacters[i])
            
        return ''.join(str(c) for c in rowcharacters[0])
